### PR TITLE
Address and gender attributes added to users.

### PR DESCRIPTION
**Pull request by Stefano-loop**
#### :link: Board reference:

* [As a user I should be able to sign up](https://loopstudio.atlassian.net/jira/software/projects/STEF/boards/42?selectedIssue=STEF-2)

---

#### Description:

Sign up feature fully functional with the following changes.  

**Address and gender attributes added to users:** 
- **Address**: Plain text with no validation whatsoever. It's a non-mandatory field.
- **Gender**: An enum with three possible values, 'male', 'female and 'other'. It's a mandatory field.

**New validations added to the user model:**
- **Password length and format validation**: Passwords must have at least 8 characters. It must contain at least one letter and one digit.
- **Gender presence and value**: Gender it's required and it needs to be within the contemplated key values in the enum.

---

#### :heavy_check_mark:Tasks:

- Added gender and address as permitted parameters in registrations controller.

- A single migration created for both new attributes (gender and address).

- Password length and format, as well as gender presence and value validations were added to the user model.

- Tests were changed to contemplate gender and address.

---

### CODE REVIEW

File: spec/factories/users.rb [hunk 1]
```diff
@@ -25,13 +25,17 @@
...
+#  address                :string
 #
 
 FactoryBot.define do
   factory :user do
     first_name { Faker::Name.first_name }
     last_name { Faker::Name.last_name }
     email { Faker::Internet.email }
+    gender { 'male' }
+    address { 'Fake St. 123' }
```
Comment (mconiglio) [inline spec/factories/users.rb]: you can use faker for this

File: spec/factories/users.rb [hunk 2]
```diff
@@ -25,13 +25,17 @@
...
+#  gender                 :enum
+#  address                :string
 #
 
 FactoryBot.define do
   factory :user do
     first_name { Faker::Name.first_name }
     last_name { Faker::Name.last_name }
     email { Faker::Internet.email }
+    gender { 'male' }
```
Comment (mconiglio) [inline spec/factories/users.rb]: maybe you can sample this `%w(male female other).sample`

File: db/structure.sql [hunk 3]
```diff
@@ -0,0 +1,171 @@
+SET statement_timeout = 0;
```
Comment (mconiglio) [inline db/structure.sql]: is this file needed? 🤔

File: app/models/user.rb [hunk 4]
```diff
@@ -25,14 +25,25 @@
...
 #  tokens                 :json
 #  must_change_password   :boolean          default(FALSE)
+#  gender                 :enum
+#  address                :string
 #
 
 class User < ApplicationRecord
   devise :database_authenticatable, :recoverable,
-         :trackable, :validatable, :registerable
+         :trackable, :validatable, :registerable, password_length: 8..128
```
Comment (mconiglio) [inline app/models/user.rb]: I think you can configure all this in the devise initializer

File: app/models/user.rb [hunk 5]
```diff
@@ -25,14 +25,25 @@
...
+  PASSWORD_FORMAT = /\A(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]+)\z/
+  PASSWORD_FORMAT_ERROR_MESSAGE = "Password format not valid. It must contain at least one letter and digit."
+  
+  enum gender: { male: 'male', female: 'female', other: 'other' }
+
+  validates :password, format: { with: PASSWORD_FORMAT, message: PASSWORD_FORMAT_ERROR_MESSAGE }, on: :create
+  validates :password, allow_nil: true, format: { with: PASSWORD_FORMAT, message: PASSWORD_FORMAT_ERROR_MESSAGE }, on: :update
+
+  validates :first_name, :last_name, :gender, presence: true
+  validates :gender, inclusion: { in: genders.keys, message: "Gender not allowed"}
```
Comment (mconiglio) [inline app/models/user.rb]: use single quotes and add a space before the last bracket

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 6]
```diff
@@ -7,12 +7,16 @@
   let(:first_name) { 'Obi Wan' }
   let(:last_name) { 'Kenobi' }
   let(:email) { 'obikenobi@rebel.com' }
+  let(:gender) { 'male' }
+  let(:address) { 'Fake St. 123' }
```
Comment (diebas) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: you can use faker in tests also if you want

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 7]
```diff
@@ -97,7 +103,17 @@
 
       it 'returns an error message' do
         expect(json(subject)).to eq(
-          errors: [{ password: ["can't be blank"] }]
+          errors: [{ password: ["can't be blank", "Password format not valid. It must contain at least one letter and digit."] }]
+        )
+      end
+    end
+
+    context 'when password format is incorrect' do
```
Comment (diebas) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: use full sentences 'when the password format is incorrect'

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 8]
```diff
@@ -97,7 +103,17 @@
...
+      let(:password) { '12345678' }
+
+      it 'returns format error message' do
```
Comment (diebas) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: same here: 'returns a format error message'

File: app/models/user.rb [hunk 9]
```diff
@@ -25,14 +25,25 @@
...
 end
```
Comment (diebas) [inline app/models/user.rb]: it's recommended to add a blank line at the end of each file, you can set up your IDE to do it automatically while saving

File: db/structure.sql (hunk 3 above)
Comment (diebas) [inline db/structure.sql]: in my opinion this change is way to big for just adding an anum type, maybe it's not worth it, what do you think? however it's a good practice to study different ways of doing things, good for you!

Comment (Stefano-loop) [reply db/structure.sql]: **structure.sql** is automatically generated when migrations are executed, it’s equivalent to **schema.rb.** I explained briefly in the PR notes why I used it. Anyway as Diego says maybe it's too big of a change for just adding an enum type to the database.

File: spec/factories/users.rb [hunk 10]
```diff
@@ -25,13 +25,17 @@
...
+#  address                :string
+#  gender                 :integer          not null
 #
 
 FactoryBot.define do
   factory :user do
     first_name { Faker::Name.first_name }
     last_name { Faker::Name.last_name }
     email { Faker::Internet.email }
+    gender { Faker::Gender.binary_type.downcase } #se puede hacer esto de alguna forma en la clase.
```
Comment (mconiglio) [inline spec/factories/users.rb]: this only works for male and female, you don't have `other` here. also, remove the comment

File: app/models/user.rb [hunk 11]
```diff
@@ -34,5 +36,20 @@ class User < ApplicationRecord
   include DeviseTokenAuth::Concerns::User
   serialize :tokens
 
-  validates :first_name, :last_name, presence: true
+  PASSWORD_FORMAT = /\A(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]+)\z/.freeze
```
Comment (letiesperon) [inline app/models/user.rb]: This belongs to a validator

File: app/models/user.rb [hunk 12]
```diff
@@ -34,5 +36,20 @@ class User < ApplicationRecord
...
+  PASSWORD_FORMAT_ERROR_MESSAGE = 'Password format not valid.'.freeze
```
Comment (letiesperon) [inline app/models/user.rb]: Internationalize all messages visible to the final user

File: app/models/user.rb (hunk 9 above)
Comment (letiesperon) [reply app/models/user.rb]: I don't think you need to override the message. If you do, internationalize it

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 13]
```diff
@@ -97,7 +103,17 @@
...
-          errors: [{ password: ["can't be blank"] }]
+          errors: [{ password: ["can't be blank", 'Password format not valid.'] }]
+        )
+      end
+    end
+
+    context 'when password format is incorrect' do
+      let(:password) { '12345678' }
+
+      it 'returns format error message' do
```
Comment (letiesperon) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: same here. full sentences. don't omit the article

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 14]
```diff
@@ -97,7 +103,17 @@
...
+        expect(json(subject)).to eq(
+          errors: [{ password: ['Password format not valid.'] }]
```
Comment (letiesperon) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: just check that errors in password is not nil. The exact message is not needed (if you then use internationalization the tests might fail)

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 15]
```diff
@@ -97,7 +103,17 @@
 
       it 'returns an error message' do
         expect(json(subject)).to eq(
-          errors: [{ password: ["can't be blank"] }]
+          errors: [{ password: ["can't be blank", 'Password format not valid.'] }]
```
Comment (letiesperon) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: just say `format not valid`. So that when devise build the full error messages they are readable

File: Gemfile [hunk 16]
```diff
@@ -34,6 +34,7 @@ group :test do
   gem 'shoulda-matchers', '~> 4.1.0'
   gem 'simplecov', '~> 0.13.0'
   gem 'webmock', '~> 3.6.0'
+  gem 'rspec-json_expectations'
```
Comment (letiesperon) [inline Gemfile]: always put gem versions
search for the latest formally released version in rubygems

File: spec/requests/api/v1/auth/passwords/create_spec.rb [hunk 17]
```diff
@@ -1,21 +1,26 @@
...
     post api_v1_user_password_path, params: params, as: :json
-    response
   end
 
   let(:user) { create(:user) }
   let(:params) { { email: user.email } }
 
   context 'when the user exists' do
-    it { is_expected.to have_http_status(:no_content) }
+    specify do
```
Comment (mconiglio) [inline spec/requests/api/v1/auth/passwords/create_spec.rb]: why specify instead of `it`?

File: spec/requests/api/v1/auth/passwords/update_spec.rb [hunk 18]
```diff
@@ -15,16 +14,22 @@
     let(:headers) { auth_headers(user) }
 
     context 'with valid params' do
-      it { is_expected.to have_http_status(:no_content) }
+      specify do
```
Comment (mconiglio) [inline spec/requests/api/v1/auth/passwords/update_spec.rb]: same

File: spec/requests/api/v1/auth/registrations/create_spec.rb [hunk 19]
```diff
@@ -1,44 +1,51 @@
...
+      address: address,
       password: password
     }
   end
 
   context 'with correct params given' do
     let(:created_user) { User.last }
 
-    it { is_expected.to have_http_status(:ok) }
+    specify do
```
Comment (mconiglio) [inline spec/requests/api/v1/auth/registrations/create_spec.rb]: same

File: spec/requests/api/v1/auth/passwords/create_spec.rb (hunk 17 above)
Comment (letiesperon) [reply spec/requests/api/v1/auth/passwords/create_spec.rb]: https://github.com/rubocop-hq/rspec-style-guide#it-and-specify

Comment (mconiglio) [reply spec/requests/api/v1/auth/passwords/create_spec.rb]: I didn't have that (no la tenía esa) 👌

**Comment by letiesperon** [comment]
Always remember to delete the branches after merged (or setup the repository to do it automatically) @Stefano-loop
